impl/client/src/main/javascript/web/util/_a11y.js

```diff
@@ -35,6 +35,421 @@ define(["./_focus"], function(focusUtil) {
+      if (!multiselectable) {
+        element.setAttribute('aria-selected', 'true');
+      }
+      element.classList.add('keyboard-navigation-focus');
+      listboxNode.setAttribute('aria-activedescendant', element.id);
+      activeDescendant = element.id;
+
+      handleFocusChange(element);
+    }
+
+    /**
+     * @description
+     *  Defocus the specified item
+     * @param element
+     *  The element to defocus
+     */
+    function defocusItem(element) {
+      if (!element) {
+        return;
+      }
+      if (!multiselectable) {
+        element.removeAttribute('aria-selected');
+      }
+      element.classList.remove('keyboard-navigation-focus');
+    }
+
+    function clearKeysSoFarAfterDelay() {
+      if (keyClear) {
+        clearTimeout(keyClear);
+        keyClear = null;
+      }
+      var keyClear = setTimeout(
+          function () {
+            keysSoFar = '';
+            keyClear = null;
+          }.bind(this),
+          500
+      );
+    }
+
+    function findItemToFocus(key) {
+      var itemList = listboxNode.querySelectorAll('[role="option"]');
+      var character = String.fromCharCode(key);
+      var searchIndex = 0;
+
+      if (!keysSoFar) {
+        for (var i = 0; i < itemList.length; i++) {
+          if (itemList[i].getAttribute('id') == activeDescendant) {
+            searchIndex = i;
+          }
+        }
+      }
+      keysSoFar += character;
+      clearKeysSoFarAfterDelay();
+
+      var nextMatch = findMatchInRange(
+          itemList,
+          searchIndex + 1,
+          itemList.length
+      );
+      if (!nextMatch) {
+        nextMatch = findMatchInRange(itemList, 0, searchIndex);
+      }
+      return nextMatch;
+    }
+
+    function findMatchInRange(
+        list,
+        startIndex,
+        endIndex
+    ) {
+      // Find the first item starting with the keysSoFar substring, searching in
+      // the specified range of items
+      for (var n = startIndex; n < endIndex; n++) {
+        var label = list[n].innerText;
+        if (label && label.toUpperCase().indexOf(keysSoFar) === 0) {
+          return list[n];
+        }
+      }
+      return null;
+    }
+
+    /* Return the previous listbox option, if it exists; otherwise, returns null */
+    function findPreviousOption(currentOption) {
+      var allOptions = Array.prototype.slice.call(
+          listboxNode.querySelectorAll('[role="option"]')
+      ); // get options array
+      var currentOptionIndex = allOptions.indexOf(currentOption);
+      var previousOption = null;
+
+      if (currentOptionIndex > -1 && currentOptionIndex > 0) {
+        previousOption = allOptions[currentOptionIndex - 1];
+      }
+
+      return previousOption;
+    }
+
+    /* Return the next listbox option, if it exists; otherwise, returns null */
+    function findNextOption(currentOption) {
+      var allOptions = Array.prototype.slice.call(
+          listboxNode.querySelectorAll('[role="option"]')
```

Comment:
I've seen this repeated multiple times. Sure could use a `getOptions()` method...
Also, a `getOptionIndex(option)` appears useful (like `getElementIndex` below???).